src/lib.rs

```diff
@@ -2727,6 +2832,147 @@ impl error::Error for FromPathError {
+///     .expect_err("non-Unicode path failed");
+/// # #[cfg(unix)]
+/// assert_eq!(err.as_os_str(), &non_unicode_string);
+/// # #[cfg(unix)]
+/// assert_eq!(err.into_os_string(), non_unicode_string);
+/// # }
+/// ```
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct FromOsStringError {
+    os_string: OsString,
+    error: FromOsStrError,
+}
+
+impl FromOsStringError {
+    /// Returns the [`OsStr`] slice that was attempted to be converted to [`Utf8PathBuf`].
+    #[inline]
+    pub fn as_os_str(&self) -> &OsStr {
+        &self.os_string
+    }
+
+    /// Returns the [`OsString`] that was attempted to be converted to [`Utf8PathBuf`].
+    #[inline]
+    pub fn into_os_string(self) -> OsString {
+        self.os_string
+    }
+
+    /// Fetches a [`FromOsStrError`] for more about the conversion failure.
+    ///
+    /// At the moment this struct does not contain any additional information, but is provided for
+    /// completeness.
+    #[inline]
+    pub fn from_os_str_error(&self) -> FromOsStrError {
+        self.error
+    }
+
+    /// Converts self into a [`std::io::Error`] with kind
+    /// [`InvalidData`](io::ErrorKind::InvalidData).
+    ///
+    /// Many users of [`FromOsStringError`] will want to convert it into an [`io::Error`].
+    /// This is a convenience method to do that.
+    pub fn into_io_error(self) -> io::Error {
+        // NOTE: we don't currently implement `From<FromOsStringError> for io::Error`
+        // because we want to ensure the user actually desires that conversion.
+        io::Error::new(io::ErrorKind::InvalidData, self)
+    }
+}
+
+impl fmt::Display for FromOsStringError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(
+            f,
+            "OsString contains invalid UTF-8: {}",
+            // self.os_string.display() // this item is stable since `1.87.0`
+            PathBuf::from(&self.os_string).display() // msrv hack
+        )
+    }
+}
+
+impl error::Error for FromOsStringError {
+    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
+        Some(&self.error)
+    }
+}
+
+/// A possible error value while converting a [`OsStr`] to a [`Utf8Path`].
+///
+/// Produced by the `TryFrom<&OsStr>` implementation for [`&Utf8Path`](Utf8Path).
+///
+///
+/// # Examples
+///
+/// ```
+/// use camino::{Utf8Path, FromOsStrError};
+/// use std::convert::{TryFrom, TryInto};
+/// use std::ffi::OsStr;
+/// # #[cfg(unix)]
+/// use std::os::unix::ffi::OsStrExt;
+///
+/// let unicode_str = OsStr::new("/valid/unicode");
+/// let utf8_path: &Utf8Path = unicode_str.try_into().expect("valid Unicode path succeeded");
+///
+/// // Paths on Unix can be non-UTF-8.
+/// # #[cfg(unix)]
+/// let non_unicode_str = OsStr::from_bytes(b"\xFF\xFF\xFF");
+/// # #[cfg(unix)]
+/// let err: FromOsStrError = <&Utf8Path>::try_from(non_unicode_str)
+///     .expect_err("non-Unicode path failed");
+/// ```
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct FromOsStrError(());
```

Comment:
I've fixed this here, but FYI it's generally preferred to not allow types to be created externally for backwards compatibility reasons.